aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -859,6 +860,99 @@ std::pair<Tensor, Tensor> _not_zero_mask_to_col_row_indices(
+  auto layout_from = self.layout();
+  auto layout_to = layout.value_or(kSparse);
+
+  if (layout_to != kSparse && layout_to != kSparseCsr && layout_to != kSparseCsc && layout_to != kSparseBsr && layout_to != kSparseBsc) {
+    AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " is not supported");
+  }
+
+  if (layout_from == kSparse && layout_to != kSparse) {
+    if (self.sparse_dim() != 2) {
+      AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " for input tensors with sparse_dim()!=2 is not supported");
+    }
+  }
+
+  if ((layout_from == kSparseCsr && layout_to == kSparseBsr)
+      || (layout_from == kSparseCsc && layout_to == kSparseBsc)) {
+    if (self.dim() != 2 + self.dense_dim()) {
+      AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " for batched inputs is not supported");
+    }
+  }
+
+  if (blocksize.has_value()) {
+    auto blocksize_to = *blocksize;
+    if (blocksize_to[0] <= 0 || blocksize_to[1] <= 0) {
+      AT_ERROR(funcname, ": blocksize needs to be positive, but got ", blocksize_to);
+    }
+
+    if (layout_to == kSparseBsr || layout_to == kSparseBsc) {
+      if (layout_from == kSparseBsr || layout_from == kSparseBsc) {
+        auto blocksize_from = at::sparse_csr::getBlockSize(self);
+        if (!(blocksize_to == blocksize_from)) {
+          AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " with blocksize changed from ", blocksize_from, " to ", blocksize_to, " is not supported");
+        }
+      } else {
+        auto dense_dim = (layout_from == kStrided) ? dense_dim_opt.value_or(0) : self.dense_dim();
+        auto sparse_row_dim = -(dense_dim + 2);
+        auto sparse_col_dim = -(dense_dim + 1);
+        if ((self.size(sparse_row_dim) % blocksize_to[0] != 0)
+            || (self.size(sparse_col_dim) % blocksize_to[1] != 0)) {
+            AT_ERROR(funcname, ": tensor sparse size (", self.size(sparse_row_dim), ",", self.size(sparse_row_dim), ") must be divisible by given blocksize (", blocksize_to[0], ",", blocksize_to[1], ")");
+        }
+      }
+    } else {
+      AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " with blocksize argument given is not supported");
+    }
+  } else {
+    if (layout_to == kSparseBsr || layout_to == kSparseBsc) {
+      AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " without blocksize argument given is not supported");
+    }
+  }
+
+  if (dense_dim_opt.has_value()) {
+    if (layout_from != kStrided) {
+      AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " with dense_dim argument given is not supported");
+    }
+
+    auto dense_dim = *dense_dim_opt;
+    if (layout_to == kSparse) {
+      if (dense_dim < 0 || dense_dim > self.dim()) {
```

Comment:
Fixed.

Reply:
Actually, fixed other parts of the code so that for example:
```python
dense = torch.eye(2)
print(dense.to_sparse(layout=torch.sparse_coo, dense_dim=2))
```
works now, so `dense_dim` could be `self.dim()` here.

Reply:
Is `dense.to_sparse(0)` also working? Its behavior (success or exception) ought to be the same as the behavior of `dense.to_sparse(layout=torch.sparse_coo, dense_dim=dense.dim())`.

AFAIK, we don't have the concept of scalar sparse tensor and tests for it. This is an important new feature that should be discussed/introduced separately from this PR. What do you think?

^ @cpuhrsch @amjames

Reply:
That's right, reverted.